### DISCUSSION TITLE
Smart design for multiple React Modals

### DISCUSSION THREAD

**Discussion by GonzalesMatthew** [nss-evening-cohort-14/evening-client-side#283] [Help Me!] [answered]
Matthew Gonzales

I haven't seen anything related to this in the help tickets.

I'm using Modal reactstrap (because it looks great) both for adding and updating my tasks. The thing is I have multiple modals now (one for adding, and one per task)! This is a problem because if I toggle one modal... I toggle them all! Therefore the last modal sits on top of a stack of modals like some sickening virtual deck of cards. 

My investigation has led me to a handful of topics to research:
(1) I've tried developing logic to only target one Modal using id's, but my Modal gets called in separate places in my project ... so perhaps I'm just adding crazy on top of crazy here. See my flowchart for what I mean, specifically where ModalContainer.js is rendered:

https://docs.google.com/presentation/d/16cxnh2h54LLz4B3zN4nnbWFl6mPgwnGqWR94YYEu9jc/edit#slide=id.p

(Also, links to my src code where I use ModalContainer.js so you have a clearer picture):
- Tasks, where my tasks call up ModalContainer.js for updating:
https://github.com/GonzalesMatthew/CAPSTONE-matts-music-monitor/blob/main/src/components/Task.js
- UserLanding, where I add new tasks:
https://github.com/GonzalesMatthew/CAPSTONE-matts-music-monitor/blob/main/src/components/UserLanding.js

I was thinking of something like data-target, but I no longer think that is a smart thing to attempt in React. I've punted this line of thinking for now...

(2) Maybe I want to reconfigure my frenzied Modal creation into, perhaps a single Modal, but then I have to figure out how to pass additional props depending on where I am in my app to that one Modal. Again reference the above flowchart for where I need ModalContainer.js and what I need from it.

Once again, this doesn't quite seem like IT! Closer... but... with React I feel like there is a smarter design on the horizon.

(3) I've discovered this promising concept of Portal which will let me render ModalContainer.js elsewhere in my app, not just from its parent Component per usual. 

https://www.reddit.com/r/reactjs/comments/blww7z/best_design_pattern_for_handling_multiple_modals/
https://reactjs.org/docs/portals.html
https://blog.logrocket.com/learn-react-portals-by-example/

So I am going to further investigate this line of thinking. If possible, I can call my Modal from elsewhere in my app and pass it the different props it needs for updating. This definitely feels like the React way! (This is the way.) However I haven't reached the end of this research topic yet so don't know.

Conclusion:
I'm afraid my research may prove fruitless, or that I'm entirely missing the point of React in this scenario - very me thing to do. If someone has a smart design for Modal here, either multiple Modals, or one magical Modal that accepts different props from different locations in my app, I would love to hear your ideas!

Thank you for reading!

**Comment by TrinityChristiana** [comment, marked as answer]
Looks like both modals are [looking at the same modal state here](https://github.com/GonzalesMatthew/CAPSTONE-matts-music-monitor/blob/main/src/views/UserProfile.js#L10) 

So when `modalStatus` is true, both modals see that and open. When it’s false, both modals see that and close. Try to find a way to separate this out so each modal is looking at a different state.

**Reply by GonzalesMatthew** [reply]
I was noticing that too, but wouldn't I have to create a hook for each modal? I have a modal per submitted task, so as I add tasks I add modals. I was getting confused how to make new hooks for each new modal automatically that didn't cross-reference the other ones?

**Reply by GonzalesMatthew** [reply]
Hey all,
Trinity helped me with the solution so let me tell you what we had to do to get this working correctly.

Problem: 
I defined modal status (false,true) and the toggle function too early, and used it for all my modals (add-modal, and each update-modal that was created). I passed these props down through my app into both locations where ModalContainer.js gets called. Big mistake. (Did someone say steak?)
Solution:
Define a separate set of status and toggle functions for both places where I call my ModalContainer.js (once for the Add button, and once for the Component where I map my task objects, which each creates its own modal.
The add set:
```
const [modal, setModal] = useState(false);
const toggle = () => {
  setModal(!modal);
}
```
The update set:
```
const [modalTwo, setModalTwo] = useState(false);
const toggleTwo = () => {
  setModalTwo(!modalTwo);
}
```
The finer detail is to only define these sets in the same place you call your Modal. This way when you pass this props into your Modal you are therefore creating separate instances. And furthermore, now that you've done this, each instance of your update-version of your Modal will indeed be its own instance, and will only apply to that instance. So even though they may share the same hook names and values, React knows to only apply them to that instance of the update-version. :)
I hope this makes sense, but take a look at my code here for details:
- Add version: line 38
 https://github.com/GonzalesMatthew/CAPSTONE-matts-music-monitor/blob/main/src/components/UserLanding.js
- Update version: line 36
 https://github.com/GonzalesMatthew/CAPSTONE-matts-music-monitor/blob/main/src/components/Task.js

Hope this helps anyone doing something similar

**Reply by drteresavasquez** [reply]
Excellent!
